examples/src/main/scala/sttp/tapir/examples/OAuth2GithubHttp4sServer.scala

```diff
@@ -0,0 +1,124 @@
+package sttp.tapir.examples
+
+import cats.effect._
+import cats.syntax.all._
+import io.circe.generic.auto._
+import org.http4s.HttpRoutes
+import org.http4s.server.Router
+import org.http4s.server.blaze.BlazeServerBuilder
+import org.http4s.syntax.kleisli._
+import pdi.jwt.{JwtAlgorithm, JwtCirce, JwtClaim}
+import sttp.client3._
+import sttp.client3.asynchttpclient.cats.AsyncHttpClientCatsBackend
+import sttp.model.StatusCode
+import sttp.tapir._
+import sttp.tapir.generic.auto._
+import sttp.tapir.json.circe._
+import sttp.tapir.server.http4s.Http4sServerInterpreter
+
+import java.time.Instant
+import scala.collection.immutable.ListMap
+import scala.concurrent.ExecutionContext
+
+object OAuth2GithubHttp4sServer extends App {
+
+  implicit val ec: ExecutionContext = scala.concurrent.ExecutionContext.Implicits.global
+  implicit val contextShift: ContextShift[IO] = IO.contextShift(ec)
+  implicit val timer: Timer[IO] = IO.timer(ec)
+  implicit val ce: ConcurrentEffect[IO] = IO.ioConcurrentEffect
+
+  // github application details
+  val clientId = "<put your client id here>"
+  val clientSecret = "<put your client secret>"
+
+  // algorithm used for jwt encoding and decoding
+  val jwtAlgo = JwtAlgorithm.HS256
+  val jwtKey = "my secret key"
+
+  type Limit = Int
+  type AuthToken = String
+
+  case class AccessDetails(token: String)
+
+  val authorizationUrl = "https://github.com/login/oauth/authorize"
+  val accessTokenUrl = "https://github.com/login/oauth/access_token"
+
+  val authOAuth2 = auth.oauth2.authorizationCode(
+    authorizationUrl,
+    accessTokenUrl,
+    ListMap.empty)
+
+  // endpoint declarations
+  val login: Endpoint[Unit, Unit, String, Any] =
+    endpoint.get.in("login")
+      .out(statusCode(StatusCode.PermanentRedirect))
+      .out(header[String]("Location"))
+
+  val loginGithub: Endpoint[String, String, AccessDetails, Any] =
+    endpoint.get.in("login" / "oauth2" / "github")
+      .in(query[String]("code"))
+      .out(jsonBody[AccessDetails])
+      .errorOut(stringBody)
+
+  val secretPlace: Endpoint[String, String, String, Any] =
+    endpoint
+      .get
+      .in("secret-place")
+      .in(authOAuth2)
+      .out(stringBody)
+      .errorOut(stringBody)
+
+  // converting endpoints to routes
+
+  // simply redirect to github auth service
+  val loginRoute: HttpRoutes[IO] = Http4sServerInterpreter.toRoutes(login)(
+    _ => IO(s"$authorizationUrl?client_id=$clientId".asRight[Unit]))
+
+  // after successful authorization github redirects you here
+  val loginGithubRoute: HttpRoutes[IO] = Http4sServerInterpreter.toRoutes(loginGithub)(
+    code => AsyncHttpClientCatsBackend[IO]().flatMap(backend => {
```

Comment:
It would be good to create the client only once, not on every request. `AsyncHttpClientCatsBackend[IO]()` allocates a thread pool, so a backend should be handled as a resource. There's a `.resource` variant which also deallocates the backend when it's no longer used.